### PR TITLE
Rewrite Swift Example with Swift 3

### PR DESCRIPTION
**Pull request by cdzombak**
Rather the maintaining the Swift example as a rewrite of the ObjC one, I’ve changed it to demonstrate how Swift types may be boxed for use with PhotoViewer, and to demonstrate how PhotoViewer may be tied into an app’s existing data layer.

### CODE REVIEW

File: Example-Swift/ViewController.swift [hunk 1]
```diff
@@ -13,25 +13,25 @@ import NYTPhotoViewer
...
+        present(photosViewController, animated: true, completion: nil)
         
         updateImagesOnPhotosViewController(photosViewController, afterDelayWithPhotos: photos)
     }
     
-    func updateImagesOnPhotosViewController(photosViewController: NYTPhotosViewController, afterDelayWithPhotos: [ExamplePhoto]) {
+    func updateImagesOnPhotosViewController(_ photosViewController: NYTPhotosViewController, afterDelayWithPhotos: [ExamplePhoto]) {
         
-        let delayTime = dispatch_time(DISPATCH_TIME_NOW, 5 * Int64(NSEC_PER_SEC))
+        let delayTime = DispatchTime.now() + Double(5 * Int64(NSEC_PER_SEC)) / Double(NSEC_PER_SEC)
```
Comment (zadr) [inline Example-Swift/ViewController.swift]: Random drive-by note:

In Swift 3, there's an overload of `+` that lets you add a `DispatchTime`  to `Double`s, so you could write this as `DispatchTime.now() + 5.0`.

Or, there's a new enum, `DispatchTimeInterval`, that has a `seconds(Int)` case, so you could also write this as `DispatchTime.now() + DispatchTimeInterval.seconds(5)`.

Comment (cdzombak) [inline Example-Swift/ViewController.swift]: I didn't know about that! Thanks!

File: Example-Swift/Photo.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//
+//  Created by Chris Dzombak on 2/2/17.
+//  Copyright © 2017 NYTimes. All rights reserved.
+//
+
+import UIKit
+
+/// A photo may often be represented in a Swift application as a value type.
+struct Photo {
+    // Ths would usually be a URL, but for this demo we load images from the bundle.
```
Comment (Veronique) [inline Example-Swift/Photo.swift]: `This would...`

File: Example-Swift/Photo.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    let name: String
+
+    let summary: String
+    let credit: String
+
+    let identifier: Int
+}
```
Comment (Veronique) [inline Example-Swift/Photo.swift]: Please remove the extra lines between these.

File: Example-Swift/PhotoBox.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    var image: UIImage?
+    var imageData: Data?
+    var placeholderImage: UIImage?
+
+    var attributedCaptionTitle: NSAttributedString?
+
+    var attributedCaptionSummary: NSAttributedString? {
+        let attributes = [NSForegroundColorAttributeName: UIColor.white,
+                          NSFontAttributeName: UIFont.preferredFont(forTextStyle: .body)]
+        return NSAttributedString(string: value.summary, attributes: attributes)
```
Comment (Veronique) [inline Example-Swift/PhotoBox.swift]: I know it is just an example but any reason why you changed the text color?

File: Example-Swift/PhotoViewerCoordinator.swift [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+//  PhotoViewerCoordinator.swift
+//  NYTPhotoViewer
+//
+//  Created by Chris Dzombak on 2/2/17.
+//  Copyright © 2017 NYTimes. All rights reserved.
+//
+
+import NYTPhotoViewer
+
+/// Coordinates interaction between the application's data layer and the photo viewer component.
```
Comment (Veronique) [inline Example-Swift/PhotoViewerCoordinator.swift]: `application data layer`

File: Example-Swift/ViewController.swift [hunk 6]
```diff
@@ -9,102 +9,69 @@
...
 
             shareActivityViewController.popoverPresentationController?.barButtonItem = photosViewController.rightBarButtonItem
-            photosViewController.presentViewController(shareActivityViewController, animated: true, completion: nil)
+            photosViewController.present(shareActivityViewController, animated: true, completion: nil)
 
             return true
         }
         
         return false
     }
```
Comment (Veronique) [inline Example-Swift/ViewController.swift]: Code style suggestion:
`
if UIDevice.current.userInterfaceIdiom != .pad { return false }
guard let photoImage = photo.image else { return false }
`
No need for another return after that.

File: Example-Swift/ViewController.swift [hunk 7]
```diff
@@ -9,102 +9,69 @@
...
-            label.backgroundColor = UIColor.redColor()
-            return label
+    func photosViewController(_ photosViewController: NYTPhotosViewController, referenceViewFor photo: NYTPhoto) -> UIView? {
+        guard let box = photo as? NYTPhotoBox else { return nil }
+
+        if box.value.name == ReferencePhotoName {
+            return imageButton
         }
+
         return nil
```
Comment (Veronique) [inline Example-Swift/ViewController.swift]:
```
return (box.value.name == ReferencePhotoName) ? imageButton : nil
```

**Comment by Veronique** [comment]
@cdzombak This example crashes when attempting to swipe to the second photo. I believe it is an issue with the indexing of the photos.

File: Example-Swift/PhotoViewerCoordinator.swift [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    @objc
+    func index(of photo: NYTPhoto) -> Int {
+        guard let box = photo as? NYTPhotoBox else { return NSNotFound }
+        return slideshow.index(where: { $0.value.identifier == box.value.identifier }) ?? NSNotFound
+    }
+
+    @objc
+    func photo(at index: Int) -> NYTPhoto? {
+        guard index < slideshow.count else { return nil }
+        guard index >= 0 else {
```
Comment (Veronique) [inline Example-Swift/PhotoViewerCoordinator.swift]: This second guard shouldn't be needed.

File: Example-Swift/PhotosProvider.swift [hunk 9]
```diff
@@ -7,39 +7,43 @@
...
+
+extension PhotosProvider {
+    fileprivate func demoPhoto(identifier: Int) -> Photo {
+        let photoName: String
+        let photoSummary: String
+        let photoCredit: String
+
+        if (arc4random_uniform(2) == 0 && identifier != 0) {
+            photoName = "Chess"
+            photoSummary = "Children gather around a game of chess during the Ann Arbor Summer Festival. (Photo ID \(identifier))"
```
Comment (Veronique) [inline Example-Swift/PhotosProvider.swift]: In the original example, the photo index was used to populate the caption. Can you do the same here to be consistent? Thanks.

File: Example-Swift/PhotoBox.swift (hunk 4 above)
Comment (cdzombak) [inline Example-Swift/PhotoBox.swift]: Not really. I think this looks a little nicer, particularly since this demo doesn't have titles on its captions.

**Comment by cdzombak** [comment]
@Veronique That crash was resolved by #227, which I just merged into this branch in e66be76.

File: Example-Swift/Photo.swift (hunk 2 above)
Comment (cdzombak) [inline Example-Swift/Photo.swift]: Fixed in https://github.com/NYTimes/NYTPhotoViewer/pull/223/commits/141e05bb820ccb27daf361502c2ab4a50751dfda

File: Example-Swift/Photo.swift (hunk 3 above)
Comment (cdzombak) [inline Example-Swift/Photo.swift]: Done in https://github.com/NYTimes/NYTPhotoViewer/pull/223/commits/141e05bb820ccb27daf361502c2ab4a50751dfda , except for `identifier` since it's distinct from other Photo properties

File: Example-Swift/PhotoViewerCoordinator.swift (hunk 5 above)
Comment (cdzombak) [inline Example-Swift/PhotoViewerCoordinator.swift]: I think either phrasing is correct.

File: Example-Swift/ViewController.swift (hunk 6 above)
Comment (cdzombak) [inline Example-Swift/ViewController.swift]: Done in https://github.com/NYTimes/NYTPhotoViewer/pull/223/commits/141e05bb820ccb27daf361502c2ab4a50751dfda

File: Example-Swift/ViewController.swift (hunk 7 above)
Comment (cdzombak) [inline Example-Swift/ViewController.swift]: Done in https://github.com/NYTimes/NYTPhotoViewer/pull/223/commits/141e05bb820ccb27daf361502c2ab4a50751dfda

File: Example-Swift/PhotoViewerCoordinator.swift (hunk 8 above)
Comment (cdzombak) [inline Example-Swift/PhotoViewerCoordinator.swift]: Done in 76fc061

File: Example-Swift/PhotosProvider.swift (hunk 9 above)
Comment (cdzombak) [inline Example-Swift/PhotosProvider.swift]: The photo index is still included here, and I'd prefer to show some more ral-looking instances in the demos for this project.

**Comment by Veronique** [comment]
@cdzombak I see an error when I run the tests on the project: in `NYTPhotosDataSourceTests`, `NYTPhotosDataSource.h` file cannot be found. I know it is not related to this example but can you fix it before merging? Thanks.

File: Example-Swift/PhotoViewerCoordinator.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                box.image = result
+                self?.photoViewer.update(box)
+            })
+        }
+    }
+
+    // MARK: NYTPhotoViewerDataSource
+
+    @objc
+    var numberOfPhotos: NSNumber? {
```
Comment (zadr) [inline Example-Swift/PhotoViewerCoordinator.swift]: Small thing, but since it's sample code that may be copied/pasted: Swift's `Int` is boxed to `NSNumber` when needed, so you could
```swift
    @objc
    var numberOfPhotos: Int? {
        return slideshow.count
    }
```

File: Example-Swift/PhotosProvider.swift [hunk 11]
```diff
@@ -7,39 +7,43 @@
...
-let PrimaryImageName = "NYTimesBuilding"
-let PlaceholderImageName = "NYTimesBuildingPlaceholder"
+/// A component of your data layer, which might load photos from the cache or network.
+final class PhotosProvider {
+    typealias Slideshow = [Photo]
 
-class PhotosProvider: NSObject {
+    /// Simulate a synchronous fetch of a slideshow, perhaps from a local database.
+    func fetchDemoSlideshow() -> Slideshow {
+        return Array(0...4).map { demoPhoto(identifier: $0) }
```
Comment (zadr) [inline Example-Swift/PhotosProvider.swift]: You don't need to wrap the range in an array initializer here; `(0...4).map { … }` will also do what you want!

File: Example-Swift/PhotoViewerCoordinator.swift (hunk 10 above)
Comment (cdzombak) [inline Example-Swift/PhotoViewerCoordinator.swift]: On Xcode 8.2.1, this leaves me with an error:

screen shot 2017-02-15 at 16 55 45

File: Example-Swift/PhotosProvider.swift (hunk 11 above)
Comment (cdzombak) [inline Example-Swift/PhotosProvider.swift]: Good catch! Fixed in 4bea18d

**Comment by Veronique** [comment]
@cdzombak Could you kindly fix the error in the tests?
> There is an error when running the tests: in NYTPhotosDataSourceTests, NYTPhotosDataSource.h file cannot be found. I know it is not related to this example but can you fix it before merging? Thanks.
